In [4]:
import pandas as pd

# Работа с Pandas

Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [5]:
googleplaystore_pro_df=pd.read_csv('C:/Users/Sergey/Downloads/googleplaystore_pro.csv')
googleplaystore_pro_df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type  Price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

Отобразите количество строк и столбцов в DataFrame

In [6]:
googleplaystore_pro_df.shape

(10841, 13)

# PRO часть (Не обязательна к выполнению, но желательна :))


Измените названия всех столбцов приведя их к нижнему регистру

Приведем названия столбцов к нижнему регистру

In [7]:
column_names=list(googleplaystore_pro_df.columns)

In [8]:
print(column_names)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [9]:
lower_column_names=[]
for word in column_names:
    lower_column_names.append(word.lower())

In [10]:
print(lower_column_names)

['app', 'category', 'rating', 'reviews', 'size', 'installs', 'type', 'price', 'content rating', 'genres', 'last updated', 'current ver', 'android ver']


In [11]:
googleplaystore_pro_df.columns=lower_column_names
googleplaystore_pro_df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content rating                     genres      last updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current ver         android ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

Выведите основные статистики по категориальным столбцам. Напишите выводы о них 

In [12]:
for column in lower_column_names:
    print(column,"_"*(15-len(column)), ":")
    print(googleplaystore_pro_df[column].describe())
    

app ____________ :
count      10841
unique      9660
top       ROBLOX
freq           9
Name: app, dtype: object
category _______ :
count      10841
unique        34
top       FAMILY
freq        1972
Name: category, dtype: object
rating _________ :
count    9367.000000
mean        4.193338
std         0.537431
min         1.000000
25%         4.000000
50%         4.300000
75%         4.500000
max        19.000000
Name: rating, dtype: float64
reviews ________ :
count     10841
unique     6002
top           0
freq        596
Name: reviews, dtype: object
size ___________ :
count                  10841
unique                   462
top       Varies with device
freq                    1695
Name: size, dtype: object
installs _______ :
count          10841
unique            22
top       1,000,000+
freq            1579
Name: installs, dtype: object
type ___________ :
count     10840
unique        3
top        Free
freq      10039
Name: type, dtype: object
price __________ :
count    10841.000000

Выводы по статистике по категориальным столбцам:
1.app ____________ :
общее кол-во заполненных строк-10841
уникальных приложений-9660
самое частотное приложение- Roblox (9поз.)

2.category _______ :
общее кол-во заполненных строк-10841
уникальных категорий-34
самая частотная категория- Family (1972поз.)

3.rating _________ :
общее кол-во заполненных строк-9367.000000 => 10841-9367=1474 пустых строк(>10%) => дополнить медианным значением рейтинга
средний рейтинг-4.193338
стандартное оклонение значений рейтинга в пределах +-0.537431
минимальный (худший) рейтинг-1.000000
25% приложений имеют рейтинг 4.000000
50% приложений имеют рейтинг 4.300000
75% приложений имеют рейтинг 4.500000
максимальный (лучший) рейтинг-19.000000 => выброс данных (max=5.0) => удалить строку/строки (проверить кол-во)

4.reviews ________ :
общее кол-во заполненных строк-10841
уникальное кол-во сделанных ревью-6002
тип данных-object => перевести в тип "float" => математические операции

5.size___________
общее кол-во заполненных строк-10841
уникальное кол-во значений размера-462
самый частотный размер-"Varies with device" (1695поз.) => выброс данных более 15% => удалять нельзя => заполним медианным значением
тип данных-object => перевести в тип "float" => математические операции

6.installs _______ :
общее кол-во заполненных строк-10841
уникальное кол-во значений установок-22
самое частотное кол-во установок- более 1,000,000 (встречается 1579 раз)
тип данных-object => перевести в тип "float" => математические операции

7.type _________ :
общее кол-во заполненных строк-10840 => 1 строка выброс данных => удалить (менее 10%) 
уникальное кол-во типов-3
самый частотный тип- Free (встречается 10039 раз)

8.price _________ :
общее кол-во заполненных строк-10841  
средняя цена-1.027273
стандартное отклонение значений цены в пределах +-15.948971
минимальная цена-0.000000
25% цен 0.000000
50% цен 0.000000
75% цен 0.000000
максимальная цена-400.000000 => выброс данных/проверить кол-во => удалить строку/строки 

9.content rating _ :
общее кол-во заполненных строк-10840 => 1 строка выброс данных => удалить (менее 10%) 
кол-во уникальных значений контент рейтинга-6
самый частотный контентный рейтинг-"Everyone"(встречается 8714 раз)

10.genres _________ :
общее кол-во заполненных строк-10841
кол-во уникальных жанров-120
самый частотный жанр-"Tools" (встречается 842 раза)

11.last updated ___ :
общее кол-во заполненных строк-10841
кол-во уникальных дат обновления-1378
самая частотная дата обновления-"August 3, 2018" (встречается 326 раз)

12.current ver ___ :
общее кол-во заполненных строк-10833 => 8 строк выброс данных => удалить (менее 10%) 
кол-во уникальных уникальных текущих версий-2832
самая частотная текущая версия-"Varies with device" (встречается 1459 раз) выброс данных более 10% => удалять нельзя => заполним медианным значением

13.android ver ___ :
общее кол-во заполненных строк-10838 => 3 строки выброс данных => удалить (менее 10%) 
кол-во уникальных уникальных версий андроид-33
самая частотная версия андроид-"4.1 and up" (встречается 2451 раз) 

Резюме: при анализе столбцов выявили выбросы данных, некоторые из них менее 10% от общей датамассы.Однако, при удалении тех или иных данных необходимо помнить,что весь датасет можно сократить максимум на 10%,не более.


Выведите на экран часть датафрейма, который соответствует условию: `приложение из категории искусства и дизайна` И `с рейтингом больше 4.5`

Найдем в столбце category категорию "Искусство и дизайн"

In [13]:
googleplaystore_pro_df['category'].value_counts()

FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
LIBRARIES_AND_DEMO       85
AUTO_AND_VEHICLES        85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
1.9                       1
Name: category, dtype: int64

In [14]:
Наша категория "ART_AND_DESIGN". Выведем на экран приложения из данной категории с рейтингом больше 4.5. 

SyntaxError: invalid syntax (3957857334.py, line 1)

In [15]:
googleplaystore_pro_df[(googleplaystore_pro_df.category=='ART_AND_DESIGN')&(googleplaystore_pro_df.rating>4.5)]

app        category  \
2     U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN   
9                         Kids Paint Free - Drawing Fun  ART_AND_DESIGN   
13                                Mandala Coloring Book  ART_AND_DESIGN   
16         Photo Designer - Write your name with shapes  ART_AND_DESIGN   
19                                         ibis Paint X  ART_AND_DESIGN   
22              Superheroes Wallpapers | 4K Backgrounds  ART_AND_DESIGN   
24                          HD Mickey Minnie Wallpapers  ART_AND_DESIGN   
25                           Harley Quinn wallpapers HD  ART_AND_DESIGN   
26                        Colorfit - Drawing & Coloring  ART_AND_DESIGN   
34                                      I Creative Idea  ART_AND_DESIGN   
36       UNICORN - Color By Number & Pixel Art Coloring  ART_AND_DESIGN   
38                       PIP Camera - PIP Collage Maker  ART_AND_DESIGN   
45    Canva: Poster, banner, card maker & graphic de...  ART_AND_DESIGN   
46    Install images with music to make video withou...  ART_AND_DESIGN   
3982                                  Cardi B Wallpaper  ART_AND_DESIGN   
4749  X Launcher: With OS11 Style Theme & Control Ce...  ART_AND_DESIGN   
4755  X Launcher Pro: PhoneX Theme, OS11 Control Center  ART_AND_DESIGN   
4759  X Launcher Pro - IOS Style Theme & Control Center  ART_AND_DESIGN   
4764     X Launcher Prime: With OS Style Theme & No Ads  ART_AND_DESIGN   
5254                            AJ Styles HD Wallpapers  ART_AND_DESIGN   
6930               Fantasy theme dark bw black building  ART_AND_DESIGN   
8888            Spring flowers theme couleurs d t space  ART_AND_DESIGN   

      rating reviews  size     installs  type  price content rating  \
2        4.7   87510  8.7M   5,000,000+  Free   0.00       Everyone   
9        4.7     121  3.1M      10,000+  Free   0.00       Everyone   
13       4.6    4326   21M     100,000+  Free   0.00       Everyone   
16       4.7    3632  5.5M     500,000+  Free   0.00       Everyone   
19       4.6  224399   31M  10,000,000+  Free   0.00       Everyone   
22       4.7    7699  4.2M     500,000+  Free   0.00   Everyone 10+   
24       4.7     118   23M      50,000+  Free   0.00       Everyone   
25       4.8     192  6.0M      10,000+  Free   0.00       Everyone   
26       4.7   20260   25M     500,000+  Free   0.00       Everyone   
34       4.7     353  4.2M      10,000+  Free   0.00           Teen   
36       4.7    8145   24M     500,000+  Free   0.00       Everyone   
38       4.7     158   11M      10,000+  Free   0.00       Everyone   
45       4.7  174531   24M  10,000,000+  Free   0.00       Everyone   
46       4.6    1070   26M     100,000+  Free   0.00       Everyone   
3982     4.8     253  3.7M      50,000+  Free   0.00       Everyone   
4749     4.7    5754  4.4M     100,000+  Free   0.00       Everyone   
4755     4.7     801  3.5M       5,000+  Paid   1.99       Everyone   
4759     4.8    1216  8.6M      10,000+  Paid   1.99       Everyone   
4764     4.7     149  3.5M       1,000+  Paid   1.99       Everyone   
5254     4.8      48   25M       5,000+  Free   0.00       Everyone   
6930     4.8      41  1.9M       5,000+  Free   0.00       Everyone   
8888     5.0       1  2.9M         100+  Free   0.00       Everyone   

                       genres       last updated current ver   android ver  
2                Art & Design     August 1, 2018       1.2.4  4.0.3 and up  
9     Art & Design;Creativity       July 3, 2018         2.8  4.0.3 and up  
13               Art & Design      June 26, 2018       1.0.4    4.4 and up  
16               Art & Design      July 31, 2018         3.1    4.1 and up  
19               Art & Design      July 30, 2018       5.5.4    4.1 and up  
22               Art & Design      July 12, 2018     2.2.6.2  4.0.3 and up  
24               Art & Design       July 7, 2018       1.1.3    4.1 and up  
25               Art & Design     April 25, 2018         1.5    3.0 and up  

Посчитатайте какое количество игр содержится в рейтинге, используя параметр разбиения (bins) равным 100. Выведите на экран топ-10 игр по количеству в рейтинге. Напишите краткий вывод.
То есть сначала отфильтровываете лишь игры из всего датафрейма, после чего применяете value_counts к столбцу с рейтингом используя параметр, который указан выше. И выводите первые 10 строк отсортированных по убыванию.

Отфильтруем игры из всего датафрейма.

In [16]:
googleplaystore_pro_df[googleplaystore_pro_df.category=='GAME']

app category  rating   reviews size  \
1653                         ROBLOX     GAME     4.5   4447388  67M   
1654                 Subway Surfers     GAME     4.5  27722264  76M   
1655               Candy Crush Saga     GAME     4.4  22426677  74M   
1656                      Solitaire     GAME     4.7    254258  23M   
1657                 Bubble Shooter     GAME     4.5    148897  46M   
...                             ...      ...     ...       ...  ...   
10791             Winter Wonderland     GAME     4.0      1287  38M   
10792        Soccer Clubs Logo Quiz     GAME     4.2     21661  16M   
10793                     Sid Story     GAME     4.4     28510  78M   
10803  Fatal Raid - No.1 Mobile FPS     GAME     4.3     56496  81M   
10804                  Poker Pro.Fr     GAME     4.2      5442  17M   

             installs  type  price content rating  \
1653     100,000,000+  Free    0.0   Everyone 10+   
1654   1,000,000,000+  Free    0.0   Everyone 10+   
1655     500,000,000+  Free    0.0       Everyone   
1656      10,000,000+  Free    0.0       Everyone   
1657      10,000,000+  Free    0.0       Everyone   
...               ...   ...    ...            ...   
10791         50,000+  Free    0.0       Everyone   
10792      1,000,000+  Free    0.0       Everyone   
10793        500,000+  Free    0.0           Teen   
10803      1,000,000+  Free    0.0           Teen   
10804        100,000+  Free    0.0           Teen   

                             genres       last updated   current ver  \
1653   Adventure;Action & Adventure      July 31, 2018  2.347.225742   
1654                         Arcade      July 12, 2018        1.90.0   
1655                         Casual       July 5, 2018     1.129.0.2   
1656                           Card     August 1, 2018       2.137.0   
1657                         Casual      July 17, 2018        1.20.1   
...                             ...                ...           ...   
10791                          Word  December 18, 2013           1.0   
10792                        Trivia       May 24, 2018        1.3.81   
10793                          Card     August 1, 2018         2.6.6   
10803                        Action     August 7, 2018       1.5.447   
10804                          Card       May 22, 2018         4.1.3   

        android ver  
1653     4.1 and up  
1654     4.1 and up  
1655     4.1 and up  
1656     4.1 and up  
1657   4.0.3 and up  
...             ...  
10791    2.2 and up  
10792    4.0 and up  
10793  4.0.3 and up  
10803    4.0 and up  
10804    2.3 and up  

[1144 rows x 13 columns]

Применим "value_counts" к столбцу с рейтингом используя параметр bin=100 и выведем первые 10 строк отсортированных диапазонов по убыванию.

In [17]:
game_df=googleplaystore_pro_df[googleplaystore_pro_df.category=='GAME']['rating'].value_counts(bins=100).reset_index().head(10)
game_df

index  rating
0  (4.48, 4.52]     174
1   (4.36, 4.4]     166
2  (4.28, 4.32]     151
3   (4.56, 4.6]     121
4   (4.16, 4.2]     115
5  (4.08, 4.12]      72
6  (4.68, 4.72]      67
7   (3.96, 4.0]      55
8   (3.76, 3.8]      41
9  (3.88, 3.92]      35

Получили диапазон рейтинга самых популярных игр (4.48, 4.52). Выведем топ 10 наиболее популярных игр по рейтингу (также добавим критерий "installs"). Предварительно "очистим" столбец "installs" от '+' и ','.

In [18]:
#отключим предупреждение "SettingWithCopyWarning message"
pd.options.mode.chained_assignment = None
top_df=googleplaystore_pro_df[(googleplaystore_pro_df.category=='GAME')&(googleplaystore_pro_df.rating>4.48)]
#переведем столбец installs из object в тип float
top1_df={}
top1_df['f'] = top_df['installs'].copy()
top1_df['f'] = top1_df['f'].str.replace(',', '', regex=True).str.replace('+', '', regex=True).astype(float)
top_df['installs_f']=top1_df['f'].copy()
top_df.sort_values(by=['rating','installs_f'], ascending=[False,False]).head(10)

app category  rating reviews  \
5822   Flippy Axe : Flip The Knife & Axe Simulator     GAME     5.0       7   
7698                                     CP Trivia     GAME     5.0       5   
10721                            Mad Dash Fo' Cash     GAME     5.0      14   
5795                              Axe Champs! Wars     GAME     5.0       8   
9056                   Santa's Monster Shootout DX     GAME     5.0       4   
7402                          Trovami se ci riesci     GAME     5.0      11   
7466                                        211:CK     GAME     5.0       8   
9265                                      EC Mover     GAME     5.0       5   
10776                             Monster Ride Pro     GAME     5.0       1   
6700                              Brick Breaker BR     GAME     5.0       7   

       size installs  type  price content rating  genres        last updated  \
5822    15M     100+  Free   0.00       Everyone  Arcade  September 27, 2017   
7698    12M     100+  Free   0.00       Everyone  Trivia      August 6, 2018   
10721   16M     100+  Free   0.00       Everyone  Arcade       June 19, 2017   
5795    25M      50+  Free   0.00       Everyone  Arcade       June 26, 2018   
9056    33M      50+  Paid   1.99           Teen  Action     August 15, 2013   
7402   6.1M      10+  Free   0.00       Everyone  Arcade      March 11, 2017   
7466    38M      10+  Paid   0.99           Teen  Arcade      April 11, 2018   
9265   4.6M      10+  Free   0.00       Everyone  Racing      August 1, 2018   
10776   24M      10+  Free   0.00       Everyone  Racing       March 5, 2018   
6700    19M       5+  Free   0.00       Everyone  Arcade       July 23, 2018   

      current ver   android ver  installs_f  
5822          1.1    4.2 and up       100.0  
7698         0.99    5.0 and up       100.0  
10721        2.5a    4.1 and up       100.0  
5795          1.1    4.1 and up        50.0  
9056         1.05    2.2 and up        50.0  
7402          0.1    2.3 and up        10.0  
7466          1.3    4.1 and up        10.0  
9265         1.11  4.0.3 and up        10.0  
10776         2.0    2.3 and up        10.0  
6700          1.0    4.1 and up         5.0

Вывод: Самые популярные игры в разрезе рейтинг/кол-во скачиваний:
1. Flippy Axe : Flip The Knife & Axe Simulator	
2. CP Trivia
3. Mad Dash Fo' Cash
4. Axe Champs! Wars
5. Santa's Monster Shootout DX	
6. Trovami se ci riesci	
7. 211:CK
8. EC Mover
9. Monster Ride Pro	
10. Brick Breaker BR	
Самые популярные жанры: аркады и гонки (arcade/racing). 80% игр бесплатны к скачиванию. Размер всех игр не превышает 40 мгб. 

Создайте функцию , которая будет применена к столбцу рейтинг методом `apply()`.
Функция должна принимать строку с рейтингом и возвращать:
 - High rating, если рейтинг больше или равен 4.5
 - Middle rating, если рейтинг между 3.8 и 4.5
 - Low rating, если рейтинг меньше 3.8 
Результат работы функции запишите в новый столбец с названием `categorical_rating`

In [19]:
def rate_eval(rating):
    if rating>=4.5:
        return "High rating"
    elif rating>3.8 and rating<4.5:
        return "Middle rating"
    elif rating<3.8:
        return "Low rating"
    else:
        return "no data"

In [20]:
googleplaystore_pro_df['rating_evaluation']=googleplaystore_pro_df['rating'].apply(rate_eval)
googleplaystore_pro_df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content rating                     genres      last updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current ver         android ver rating_evaluation  
0                   1.0.0        4.0.3 and up     Middle rating  
1                   2.0.0        4.0.3 and up     Middle rating  
2                   1.2.4        4.0.3 and up       High rating  
3      Varies with device          4.2 and up       High rating  
4                     1.1          4.4 and up     Middle rating  
...                   ...                 ...               ...  
10836                1.48          4.1 and up       High rating  
10837                 1.0          4.1 and up       High rating  
10838                 1.0          2.2 and up           no data  
10839  Varies with device  Varies with device       High rating  
10840  Varies with device  Varies with device       High rating  

[10841 rows x 14 columns]

Выведите кол-во уникальных значений в столбцах отсортированных по убыванию на экран:

In [21]:
display(googleplaystore_pro_df.value_counts())

app                                                 category          rating  reviews  size                installs      type  price  content rating  genres                       last updated        current ver                              android ver         rating_evaluation
CBS Sports App - Scores, News, Stats & Watch Live   SPORTS            4.3     91031    Varies with device  5,000,000+    Free  0.00   Everyone        Sports                       August 4, 2018      Varies with device                       5.0 and up          Middle rating        4
Nick                                                ENTERTAINMENT     4.2     123279   25M                 10,000,000+   Free  0.00   Everyone 10+    Entertainment;Music & Video  January 24, 2018    2.0.8                                    4.4 and up          Middle rating        4
Quizlet: Learn Languages & Vocab with Flashcards    EDUCATION         4.6     211856   Varies with device  10,000,000+   Free  0.00   Everyone        Educat

__Super-Hard задания__

Решите задачу, где мы создавали свою собственную функцию, но вместо функции `def` используйте `лямбда-функцию` внутри `apply()`

предыдущая функция:
def rate_eval(rating):
    if rating>=4.5:
        return "High rating"
    elif rating>3.8 and rating<4.5:
        return "Middle rating"
    elif rating<3.8:
        return "Low rating"
    else:
        return "no data"

In [22]:
googleplaystore_pro_df['rating_eval']=googleplaystore_pro_df['rating'].apply(lambda rating: "High rating" if rating>=4.5 else("Low rating" if rating<3.8 else "Middle rating")) 
googleplaystore_pro_df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content rating                     genres      last updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current ver         android ver rating_evaluation    rating_eval  
0                   1.0.0        4.0.3 and up     Middle rating  Middle rating  
1                   2.0.0        4.0.3 and up     Middle rating  Middle rating  
2                   1.2.4        4.0.3 and up       High rating    High rating  
3      Varies with device          4.2 and up       High rating    High rating  
4                     1.1          4.4 and up     Middle rating  Middle rating  
...                   ...                 ...               ...            ...  
10836                1.48          4.1 and up       High rating    High rating  
10837                 1.0          4.1 and up       High rating    High rating  
10838                 1.0          2.2 and up           no data  Middle rating  
10839  Varies with device  Varies with device       High rating    High rating  
10840  Varies with device  Varies with device       High rating    High rating  

[10841 rows x 15 columns]

Изучите функцию `transform` в документации Pandas. Примените её к датафрейму, который будет заранее сгруппирован по столбцу `category`. С помощью фукнции `transform` найдите средний `rating` по каждой категории и запишите в столбец с названием `mean_cat_rating`. Кол-во строк в получившемся новом столбец должно быть таким же, как и в исходном датафрейме

In [31]:
googleplaystore_pro_df['mean_cat_rating']=googleplaystore_pro_df.groupby(by='category')['rating'].transform('mean')
googleplaystore_pro_df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content rating                     genres      last updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current ver         android ver rating_evaluation  \
0                   1.0.0        4.0.3 and up     Middle rating   
1                   2.0.0        4.0.3 and up     Middle rating   
2                   1.2.4        4.0.3 and up       High rating   
3      Varies with device          4.2 and up       High rating   
4                     1.1          4.4 and up     Middle rating   
...                   ...                 ...               ...   
10836                1.48          4.1 and up       High rating   
10837                 1.0          4.1 and up       High rating   
10838                 1.0          2.2 and up           no data   
10839  Varies with device  Varies with device       High rating   
10840  Varies with device  Varies with device       High rating   

         rating_eval day_of_update  mean_cat_rating  
0      Middle rating             7         4.358065  
1      Middle rating            15         4.358065  
2        High rating             1         4.358065  
3        High rating             8         4.358065  
4      Middle rating            20         4.358065  
...              ...           ...              ...  
10836    High rating            25         4.192272  
10837    High rating             6         4.192272  
10838  Middle rating            20         4.189143  
10839    High rating            19         4.346067  

Создайте функцию , которая будет применена к столбцу `last updated` и возвращать из него число месяца, если число месяца вернуть из строки невозможно, то верните вместо этого строку с значением `miss_date`.\
Запишите полученный столбец в датафрейм , название столбцу дайте `day_of_update`.

Если вы столкнётесь с некоторой ошибкой , которую вам выдаст Pandas, то изучите конструкцию Python `try` \ `except`.\
Вот псевдокод , который поможет лучше разобраться с этой конструкцией:

    Воспользуемся модулем для работы с датами: pd.to_datetime

In [24]:
from datetime import datetime
def update(last_updated):
    try: # Помещаете то, что должно выполниться внутрь try, если же внутри try появляется ошибка, то условие переключается на оператор except и выполняется всё то, что находится внутри except.
        last_updated= pd.to_datetime(last_updated)
        
        return last_updated.day
    except:
        return "miss_date"

googleplaystore_pro_df['day_of_update']=googleplaystore_pro_df['last updated'].apply(update)
display(googleplaystore_pro_df)


app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content rating                     genres      last updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current ver         android ver rating_evaluation  \
0                   1.0.0        4.0.3 and up     Middle rating   
1                   2.0.0        4.0.3 and up     Middle rating   
2                   1.2.4        4.0.3 and up       High rating   
3      Varies with device          4.2 and up       High rating   
4                     1.1          4.4 and up     Middle rating   
...                   ...                 ...               ...   
10836                1.48          4.1 and up       High rating   
10837                 1.0          4.1 and up       High rating   
10838                 1.0          2.2 and up           no data   
10839  Varies with device  Varies with device       High rating   
10840  Varies with device  Varies with device       High rating   

         rating_eval day_of_update  
0      Middle rating             7  
1      Middle rating            15  
2        High rating             1  
3        High rating             8  
4      Middle rating            20  
...              ...           ...  
10836    High rating            25  
10837    High rating             6  
10838  Middle rating            20  
10839    High rating            19  
10840    High rating            25  

[10841 rows x 16 columns]